### DISCUSSION TITLE
v0.3.14

### DISCUSSION THREAD

**Discussion by patrickgold** [florisboard/florisboard#1678] [Announcements]
## Important notes

After a long development phase the completely reworked FlorisBoard is now ready for a stable 0.3.14 release. A lot of work has gone into this, I want to thank all my contributors, translators, issue maintainers and supporters of this project for the help!

If you want to give feedback on the new release please do so in the announcement discussion for this release. Please make sure to read through the important notes down below to see why some things may have reset.

**Preferences one-time reset**

Upon installing the 0.3.14 update onto the existing stable version ALL settings, custom themes and the clipboard history will be lost. This is due to the fact that the underlying file and settings structure has completely changed, themes are now proper extensions, underlying database structures have changed too. With this new format established, the future stable updates will not cause such a big reset anymore.

**Word suggestions**

This release temporary removes support for the completely buggy old word suggestions. If you still enable word suggestions the Smartbar would show a placeholder with the current word plus a number, to test out some UI layouts in the Smartbar. Please DO NOT file an issue regarding word suggestions not working, this is known and just clutters up the issue list unnecessarily. On a new word suggestion functionality will be worked on in the 0.4.0 development cycle. 

---

## Settings UI rework

- Design and structural changes
  - The Settings UI has undergone a major overhaul and has moved from a bottom-navigation to a list-based approach
  - Coloring has been updated to be more modern and to be more unified and consistent across the UI
  - Added AMOLED dark mode in addition to the dark grey dark mode
  - Add "System default (AMOLED)" option for Settings UI theme (#1539)
- Add an preview field to try out features without having to leave the Settings UI
  - Note that the preview field may be buggy when it comes to text selection, this is a known issue but likely caused by the underlying UI library logic
- Re-design Setup UI screen (#996)
- Add a redesigned subtype manager screen
  - Allow to use any language available on the device for a subtype (useful when installing custom dictionaries for languages not included in the subtype presets)
  - Allow to manually define which popup mapping should be used
  - Subtype manager and its add/edit screens are now screen rotation safe and do not loose the input state
- Implement separate keyboard height prefs for portrait/landscape (#360)
- Add devtools screen / Add Android system debug tools (#1328)

## Backup & Restore / Auto-backup (#1498)

- Add support for Auto-backup
  - FlorisBoard now supports Android's Auto-backup by default, which allows you to back-up your FlorisBoard configuration via your system backup utility (Google Drive, Seedvault, etc.)
  - Only full backups are possible, no partial ones
- Add internal Backup & Restore screen
  - Located in Settings > Advanced and allows to perform manual backups, exported/imported as ZIP archive files
  - Backup archives are cross-track compatible (stable -- beta -- debug), with minor exceptions when a feature diff is existent between said tracks
  - Manual backups can do partial backup/restore operations as well

## Clipboard logic and history changes

- Full rework of clipboard backend
  - All items are now stored in the app's database, so switching keyboards or restarting your device won't clear your history anymore
  - Add clearing primary clip also deletes last history entry (#1448)
  - Fix clipboard adding empty string silently when actually failing (#1647)
  - Fix duplicate primary clip callback adding duplicates (#1555)
  - Fix selectedText not checking for null when performing a copy (#1637, thanks @Thumbscrew)
  - Rewrite the clipboard image backend, is now compatible with more apps
- Clipboard history improvements
  - The clipboard history is not viewable anymore if your device is locked, making the contents of your history more secure (#1250)
  - Add a note on-screen if the clipboard history is disabled or empty, so any confusion around if the clipboard history is actually enabled or not should be cleared
  - Add clear clipboard history confirmation (#1606)
  - Fixed a lot of clipping issues with tap animations

## Smartbar changes

- The Smartbar is now able to show both a primary and secondary row at once, allowing to access clipboard utilities while word suggestions are showing
- The secondary row is optional and can be placed above, below or floating over the app's UI when expanded.
- Add ability to flip the side on which the quick actions access button is located (defaults to left)
- Add auto-expand for the quick actions if no suggestions can be shown (configurable in prefs)

## Theme rework / New stylesheet editor (#1513)

- A completely new and overhauled in-app stylesheet editor is now included. Its main features are:
  - Live preview within the keyboard UI of all stylesheet changes
  - Properties and values are now selected in dropdowns, thus allowing for easy discoverability of what each property/rule supports
  - Easy definition and usage of variables for re-using colors throughout the whole stylesheet
  - It is easier than ever to define custom rules for specific keys and specific states
- Add ability to change the `shape` and `font-size` of keys and other keyboard elements
- Add ability to specify hardware-accelerated `shadow-elevation`
- Implement `border-color` and `border-width` for stylesheets (#1105)

As stated above, the new stylesheets are NOT compatible with the old 0.3.13 custom themes, you must recreate them. I am very sorry for the inconvenience caused here, but this is an absolutely necessary step in improving custom themes for future releases!

## Keyboard layout additions and changes

- Add Basic Telex for Vietnamese (#978, thanks @Hayleia)
- Add Igbo layout (#1251, thanks @nd500)
- Add Rusyn Language Language Set, Layout, Popups (#1524, thanks @svvvst)
- Add Thai layout (#1226, thanks @mrdeck4rd)
- Add ur-PK in phonetic layout (#1278, thanks @mubashir-rehman)
- Add Warang Citi layout (#1274, #1288, thanks @Singkiring57)
- Change name of Rial currency symbol (﷼) (#1323, thanks @rex07)
- Fix German uppercase letter ẞ being displayed as SS (#1321)
- Fix kurmanji kurdish letter ş (#1638, thanks @GoRaN909)
- Fix redundant % symbol (#1267, thanks @Aman9das)
- Fix some character codes in ipa layout (#1444, thanks @tsiflimagas)
- Update Kurdish layout (#1410, thanks @GoRaN909)
- Update Romanian layout (#1330, thanks @SchnWalter)
- Update Urdu phonetic layout (#1619, thanks @blueberry-robin)

## Keyboard logic and UI backend rework

- The keyboard logic and UI backend has been completely reworked, as the old backend had severe problems working with the new preference logic
  - Layouts, currency sets, popup mappings and subtype presets are now using the extension scheme. In the future this allows for imports of custom layouts etc. and for building up the community sharing repository.
- Add option to select text rather than delete it via the delete key swipe left gesture (#705, #1534)
- Fix ArrayIndexOutOfBounds crash in glide typing (#801, #927, #1398)
- Fix ConcurrentModificationException crash in glide typing (#887, #1632)
- Fix StringIndexOutOfBoundsException crash (#1140, #1301, #1519, #1662)
- Fix numeric and numeric advanced popup width (#1044)
- Fix space-bar switches to full keyboard in number-only fields (#1248)

## Localization changes

- Add Japanese (ja) as available language
- Add option to specify language name display type (#1568)
- Add ability to hide language name on the space bar (#371, #1025)
- Add full support for RTL (right-to-left) UI in both the preferences and keyboard UI
- Fix parentheses for RTL languages (#422, #1624)
- Disable bundle language split for GPlay releases
  - This fixes an issue where users getting FlorisBoard from the Google Play Store could not manually
    change the language, as only the system and default language was shipped.
- Fix long titles and summaries in Settings for English (#998)
- Fix Kurdish language not changing app language (#1620, thanks @GoRaN909)
- Fix typo in strings.xml (#1445, thanks @Spiryt7)
- Remove slurs from glide dictionary (#1306, thanks @pnppl)

## Improvement of the emoji palette view (#1589)

- Add a new basic emoji palette
  - Emoji categories and order is based on Unicde 14.0
  - Emoji history now saves your recently used emojis and sorts them by newest to oldest (#268)
  - Emoji popups work properly again (#877)
- Add ability to specify preferred emoji skin tone (#132)
- Add ability to specify emoji history max size

## Other changes

- Debug builds are now additionally installable on x86/x86_64 architectures
- Raise `targetSdk` from API 30 to 31
- Fix crash with Kotlin Coroutines (#1295)
- Fix crash dialog crashing on Android 12 when pushing a notification
- Fix crash in devtools system settings viewer (#1387)

## Known issues

- When showing the keyboard UI after longer idling, it may refuse to accept touch input at first. If this happens, hide the keyboard and tap back into the text field you want to type, then it works again. This is a temporary workaround and hopefully a fix will be found for the 0.3.15/16 hotfix releases.
- Glide typing has issues recognizing long words (unknown cause atm)

---

*Commits included in this release:* https://github.com/florisboard/florisboard/compare/v0.3.13...v0.3.14

<hr /><em>This discussion was created from the release <a href='https://github.com/florisboard/florisboard/releases/tag/v0.3.14'>v0.3.14</a>.</em>

**Comment by leso-kn** [comment]
Wow! Truely and with no doubt the largest and most impressive update of an open source project I've ever seen. Congratulations on the final release after 6 months of work, very well done!

Will report any issues I run into once it's up on F-Droid (except for suggestions-related issues ;)

**Reply by patrickgold** [reply]
Thanks a lot! Just a reminder for F-Droid builds that they will take some time (4-7 days based on previous experience) to be available, provided that nothing fails in the build process.

**Comment by GoRaN9O** [comment]
**The best android keyboard ever I love this app thank you very much for developing this great full features keyboard app I think the FlorisBoard is much more better than the GBoard and it's the only alternative too. 

BiG Thanks.**

**Comment by macleod-ee** [comment]
Just to be clear — this version does not have support for autocorrect, right? Due to the rework for v4.0 on the word suggestions? Android 11/OnePlus 8T. I have installed the FF dictionary pack, set the keyboard as the spellcheck as well.

All in all, the new version is great!

**Reply by patrickgold** [reply]
No, this stable release does not contain support for word suggestions nor autocorrect, these those are somewhat dependent on each other and will both be part of my 0.4.0 dev effort.

What works though is the spell checker, which after installing the necessary dictionaries and configuring FlorisBoard as spell checker will underline incorrect words within the app's UI. Note that the spell checker can be buggy and refuse to underline anything or show empty suggestions atm. This is something which I will hopefully be able to fix in the future.

**Reply by macleod-ee** [reply]
Great to hear! I assumed the two were interconnected, this latest release is really great work. The new settings and overall snapliness are very good and am looking forward to future versions!

**Comment by Gitoffthelawn** [comment]
Excited to try this!  Thank you for all your efforts!

After all these years, Android has been without a decent keyboard.  Hopefully this will be it!
